nixos/modules/services/networking/teleport.nix

```diff
@@ -0,0 +1,234 @@
+      type = types.str;
+      default = "0.0.0.0:3023";
+      description = "Set the address to listen on for the Teleport Daemon SSH Proxy Service";
+    };
+    public_addr = mkOption {
+      type = types.str;
+      default = "";
+      description = "Set the public address/dns entry to listen on for the Teleport Daemon SSH Proxy Web Service";
+    };
+    web_listen_addr = mkOption {
+      type = types.str;
+      default = "0.0.0.0:3008";
+      description = "Set the public address to listen on for the Teleport Daemon SSH Proxy Web Service";
+    };
+    tunnel_listen_addr = mkOption {
+      type = types.str;
+      default = "0.0.0.0:3024";
+      description = "Set the address to listen on for reverse tunneling for the Teleport Daemon SSH Proxy Web Service";
+    };
+    https_key_file = mkOption {
+      type = types.str;
+      default = "";
+      description = "Set the path to the ssl key file to enable https support for the Teleport Daemon SSH Proxy Web Service";
+    };
+    https_cert_file = mkOption {
+      type = types.str;
+      default = "";
+      description = "Set the path to the ssl certificate to enable https support for the Teleport Daemon SSH Proxy Web Service";
+    };
+  };
+  
+in {
+  options = {
+    services.teleport = {
+      enable = mkEnableOption "Teleport Cluster";
+      dataDir = mkOption {
+        type = types.path;
+        default = "/var/lib/teleport";
+        description = "Directory to store Teleport Service data";
+      };
+      configText = mkOption {
+        type = types.nullOr types.lines;
+        default = null;
+        description = "If non-null, this option defines the text that is written to teleport.yml";
+      };
+      nodename = mkOption {
+        type = types.str;
+        default = "";
+        description = "Set the name of the node for the Teleport SSH Daemon";
+      };
+      auth_token = mkOption {
+        type = types.str;
+        default = "";
+        description = "The token to auth with to your teleport cluster.";
+      };
+      auth_servers = mkOption {
+        type = types.listOf types.str;
+        default = [ "" ];
+        description = "Auth servers to use in your teleport cluster.";
+      };
+      advertise_ip = mkOption {
+        type = types.str;
+        default = "";
+        description = "IP of your node to advertise to your cluster, for NAT'd setups";
+      };
+      connection_limits = {
+        max_connection = mkOption {
+          type = types.int;
+          default = 1000;
+          description = "Max amount of connections allowed to your teleport node at once.";
+        };
+        max_users = mkOption {
+          type = types.int;
+          default = 250;
+          description = "Max amount of users allowed to connect to your teleport node at once.";
+        };
+      };
+      inherit (telTypes) ssh_service auth_service proxy_service;
+    };
+  };
+  
+  config = mkIf cfg.enable {
+    systemd.services.teleport = {
```

Comment:
To join teleport node to the cluster we have to run `teleport` as `root` and it's ok to run `auth` and `proxy` as less privileged user. AFAICS we run always as root with this implementation, don't we?